### PR TITLE
Add continuous CI benchmarking using astropy-benchmarks

### PR DESCRIPTION
**Pull request by MridulS**
~This PR moves astropy benchmarks from https://github.com/astropy/astropy-benchmarks to this repo~. This closes https://github.com/astropy/astropy/issues/12767 and https://github.com/astropy/astropy-benchmarks/issues/102. This also implements https://github.com/astropy/astropy-benchmarks/issues/106 (taken from scikit-image).

~Keeping this as draft for now to debug the CI (sorry for the notification noise 😅 )~

This PR implements a github action that will run a comparative ASV benchmark run between the head commit of the PR and the main branch when a maintainer of astropy labels the PR with something like `benchmark`. These benchmarks tests are running on shared infrastructure of github actions so it won't be the most precise ones but they *should* be enough to capture a regression/speed up that significantly affects astropy (it may pop up some false positives too).

If there is a significant performance change it will be "fail" the benchmark action and also upload the result log file to the action artefacts.

Anyone with triage permission on astropy should be able to test this PR by creating and adding `benchmark` label. The name is just a suggestion feel free to change it :)

### CODE REVIEW

**Comment by pllim** [comment]
`+792,747 −1` 😱 

Is there no other way to make the diff here smaller?

**Comment by pllim** [comment]
Would storing stuff over at https://github.com/astropy/astropy-data or https://github.com/astropy/astropy-benchmarks help?

**Comment by MridulS** [comment]
> https://github.com/astropy/astropy-benchmarks help?

Yeah we can keep all the files there and just clone it in this action. I want to make this work first inside VM :)

File: .github/workflows/ci_benchmark.yml
Comment (MridulS) [inline .github/workflows/ci_benchmark.yml:75]: This needs to be updated once https://github.com/astropy/astropy-benchmarks/pull/111 is merged.

**Comment by MridulS** [comment]
@pllim this is ready for an initial review :)

**Comment by pllim** [comment]
Thanks! I'll put this in my queue and hopefully get back to you soon. `+120` is much better, thanks!

File: .github/workflows/ci_benchmark.yml
Comment (pllim) [inline .github/workflows/ci_benchmark.yml:15]: We have a lot of old forks with Actions unintentionally enabled. Also someone might come add more triggers in the future, so I would feel better if we are stricter with the `if`, maybe similar to this:

https://github.com/astropy/astropy/blob/3d5aae9d0b435c459d4e10c4958c73568eda6552/.github/workflows/ci_cron_weekly.yml#L33

File: .github/workflows/ci_benchmark.yml
Comment (pllim) [inline .github/workflows/ci_benchmark.yml:16]: This name doesn't really tell me anything useful, so either make it less cryptic or just remove it altogether?

```suggestion
```

**Comment by pllim** [comment]
Thanks! I don't grok all the ccache stuff so maybe extra pairs of dev eyes would be good.

I also added the `benchmark` label here to see how this would work in practice. At a glance, I think this workflow runs all the benchmarks instead of just those for the affected sub-packages but maybe that is okay (unless we hit free-tier duration limits) because a lot of our sub-packages are inter-related.

If this works, I think it is going to complement nicely with https://spacetelescope.github.io/bench/astropy-benchmarks/ . 👏 Thanks! 🙇‍♀️

**Comment by pllim** [comment]
So it ran to completion in less than 30 mins but I am not sure what that means. How do I see the results? Or do I only get to see results when something fails?

https://github.com/astropy/astropy/actions/runs/7762333963/job/21172571433?pr=15779

**Comment by pllim** [comment]
Ah, I see in the log that there is a `BENCHMARKS NOT SIGNIFICANTLY CHANGED`. I wonder if there is a way to make that easier to discover. 🤔

**Comment by MridulS** [comment]
> Ah, I see in the log that there is a BENCHMARKS NOT SIGNIFICANTLY CHANGED. I wonder if there is a way to make that easier to discover. 🤔

The test will fail if there is a change, and I guess in majority of the cases there shouldn't be a change, so there shouldn't be a need to go look at the test.

If the test fails you can look at the benchmark.log file which should have the information about the failure. Does it make sense to maybe just comment the log file in the github interface?

File: .github/workflows/ci_benchmark.yml
Comment (MridulS) [inline .github/workflows/ci_benchmark.yml:15]: This could both ways, we only explicitly want to run the benchmarks or we want to continuously runs benchmarks for a specific PR (once labelled). Let me know if there are preferences here.

IMO it makes sense to keep running benchmarks on new commits which are tagged as performance critical.

**Comment by eerovaher** [comment]
It sounds reasonable that the benchmarks run if a maintainer adds a label to the pull request (although it is not clear to me we need a new one when we could use the existing `Performance` label), but this must be documented in `docs/development/workflow/maintainer_workflow.rst`.

**Comment by pllim** [comment]
Once a PR is labeled with the trigger, it makes sense to re-run benchmark on new commits. If author pushes out too many commits serially (which we see a lot, unfortunately), a maintainer can step in and warn author of the implications. Worst case scenario, maintainer can close that PR as spammy, cherry-pick the relevant commits to a new PR, and take it over.

As for `Performance` vs `benchmark`: I am not sure yet. I vote we keep them separate until we are sure, to iron out edge cases.

**Comment by pllim** [comment]
Re: https://github.com/astropy/astropy/pull/15779#issuecomment-1925219965

Yes, this makes sense. I think this should be documented somewhere in https://docs.astropy.org/en/latest/development/testguide.html (or some other section in the dev guide) but let's wait till we finalize the workflow itself first.

**Comment by pllim** [comment]
`message: 'Resource not accessible by integration'`

Hmm, need a special token?

**Comment by MridulS** [comment]
> Hmm, need a special token?

Trying to make this work without any token, or probably just need to check permissions again :)

**Comment by MridulS** [comment]
This is ready for another review @pllim, this would not create a comment in this PR as I can't test on `pull_request_target` event in this PR as I don't have write access to astropy (from what I gather). I think the PR comment bits should work but maybe we can try that out in https://github.com/astropy/astropy/pull/16088. Anyway the benchmarks.log should be available in the artefacts uploads of the github action run irrespective of the comment.

If something is broken I can create a new PR :)

**Comment by pllim** [comment]
Thanks! I will aim to review next week.

File: .github/workflows/ci_benchmark.yml
Comment (pllim) [inline .github/workflows/ci_benchmark.yml:115]: I assume this upload is so people can download and inspect the log locally? If not, might want to explain.

File: .github/workflows/ci_benchmark.yml
Comment (pllim) [inline .github/workflows/ci_benchmark.yml:150]: `pull_request_target` is not an event this workflow is subscribed to, so this will never run. Might as well delete it?

```suggestion
```

File: asv.ci.conf.json
Comment (pllim) [inline asv.ci.conf.json:12]: Do we need to pin dependency versions here to make sure we're comparing apple to apple?

File: .github/workflows/ci_benchmark.yml
Comment (pllim) [inline .github/workflows/ci_benchmark.yml:8]: We probably want auto-cancel (in case someone is spam pushing to a PR) and restrict permission?
```suggestion

concurrency:
  group: ${{ github.workflow }}-${{ github.ref }}
  cancel-in-progress: true

permissions:
  contents: read

```

**Review by pllim — CHANGES_REQUESTED** [review]
Trying to really understand this, so I have some comments and suggestions that might or might not work.

Maybe this new workflow also needs to be explained in https://docs.astropy.org/en/latest/development/testguide.html but I can do that as follow-up commit to this PR later (after I understand the workflow) if you are not comfortable touching the dev docs.

p.s. You might want to rebase to pick up recent changes to `main`.

Thanks for your patience!

File: .github/workflows/ci_benchmark.yml
Comment (MridulS) [reply .github/workflows/ci_benchmark.yml:36]: huh, just tested this again and this does work with just the synchronize event. I have no idea why I went down this rabbit hole. Thanks!

File: .github/workflows/ci_benchmark.yml
Comment (MridulS) [reply .github/workflows/ci_benchmark.yml:115]: Yup, it's to make sure people can download the logs locally.

File: asv.ci.conf.json
Comment (MridulS) [reply asv.ci.conf.json:12]: We could but in this case we are just comparing 2 commits in the same envs. The env itself isn't that important.

File: .github/workflows/ci_benchmark.yml
Comment (MridulS) [reply .github/workflows/ci_benchmark.yml:150]: Sure we can remove this for now.

File: .github/workflows/ci_benchmark.yml
Comment (MridulS) [reply .github/workflows/ci_benchmark.yml:120]: I made 2 different workflows at first to make the commenting work but didn't remove this.
Should we just remove all the commenting bits from the PR? To enable comments we need to run with `pull_request_target` event but this event can't run code from the PR. It runs on the base commit of the PR.

Comment (pllim) [reply .github/workflows/ci_benchmark.yml:99]: > Should we just remove all the commenting bits from the PR? 

If we are not using it, let's remove it for now. Otherwise, it would just rot. We can always reintroduce later if needed. Thanks!

**Comment by pllim** [comment]
CI failures should go away now if it is rerun (which would naturally happen on new push event).

**Comment by pllim** [comment]
However, "read artifact" failure is something else:

```
Error: Unable to process file command 'output' successfully.
Error: Invalid format '· Discovering benchmarks'
```

Did my suggestion introduce a bug? 😱

**Comment by MridulS** [comment]
Yeah, github changed things in action and we need this hacky way of setting output env variable inside actions https://github.com/orgs/community/discussions/26288#discussioncomment-3876281 but we can just remove that part as we aren't dealing with writing comments any more.

**Review by pllim — APPROVED** [review]
Diff looks good now. I guess only one way to find out if this really works! ;)

Thanks for implementing this much needed feature!

**Comment by pllim** [comment]
You might want to rebase one more time to pick up the updated CI checks.

**Comment by MridulS** [comment]
Let's try this on #16135 once it's merged? (assuming there is a benchmark that is already catching the code blocks there :) )

**Comment by pllim** [comment]
Let's do it. Thanks!

**Comment by astrofrog** [comment]
This is great! However could we just enable this by default? Sometimes/often regressions might be introduced by unintentional changes.

**Comment by MridulS** [comment]
>However could we just enable this by default?

This will add a 25 mins job to do the CI. I'm not super enthusiastic about that.

**Comment by pllim** [comment]
Wait... nvm. Cron does not make sense. We already have https://spacetelescope.github.io/bench/astropy-benchmarks/

**Comment by eerovaher** [comment]
Where's the documentation for the benchmarking?

**Comment by pllim** [comment]
Ooops... I think I forgot to ask for one here.

**Pull request merged**
